analyzers/src/SonarAnalyzer.Common/Rules/InsteadOfAnyBase.cs

```diff
@@ -0,0 +1,122 @@
+
+    protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(existsRule, containsRule);
+
+    protected abstract bool IsSimpleEqualityCheck(TInvocationExpression node, SemanticModel model);
+    protected abstract SyntaxNode GetArgumentExpression(TInvocationExpression invocation, int index);
+    protected abstract bool AreValidOperands(string lambdaVariable, SyntaxNode first, SyntaxNode second);
+
+    protected InsteadOfAnyBase()
+    {
+        existsRule = Language.CreateDescriptor(ExistsDiagnosticId, MessageFormat);
+        containsRule = Language.CreateDescriptor(ContainsDiagnosticId, MessageFormat);
+    }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            var invocation = (TInvocationExpression)c.Node;
+
+            if (IsNameEqualTo(invocation, nameof(Enumerable.Any))
+                && Language.Syntax.HasExactlyNArguments(invocation, 1)
+                && Language.Syntax.TryGetOperands(invocation, out var left, out var right)
+                && IsCorrectCall(right, c.SemanticModel)
+                && c.SemanticModel.GetTypeInfo(left).Type is { } type)
+            {
+                if (existsTypes.Any(x => type.DerivesFrom(x)))
+                {
+                    RaiseIssue(c, invocation, existsRule, "Exists");
+                }
+                else if (containsTypes.Any(x => type.DerivesFrom(x) && IsSimpleEqualityCheck(invocation, c.SemanticModel)))
+                {
+                    RaiseIssue(c, invocation, containsRule, "Contains");
+                }
+                else if (type.DerivesFrom(KnownType.System_Collections_Generic_List_T))
+                {
+                    if (IsSimpleEqualityCheck(invocation, c.SemanticModel))
+                    {
+                        RaiseIssue(c, invocation, containsRule, "Contains");
+                    }
+                    else
+                    {
+                        RaiseIssue(c, invocation, existsRule, "Exists");
+                    }
+                }
+            }
+        }, Language.SyntaxKind.InvocationExpression);
+
+    protected bool IsNameEqualTo(SyntaxNode node, string name) =>
+        Language.GetName(node).Equals(name, Language.NameComparison);
+
+    protected static bool IsValueTypeOrString(SyntaxNode expression, SemanticModel model) =>
+        model.GetTypeInfo(expression).Type is { } type
+        && (type.IsValueType || type.Is(KnownType.System_String));
+
+    protected bool HasValidInvocationOperands(TInvocationExpression invocation, string lambdaVariableName, SemanticModel model)
+    {
+        if (IsNameEqualTo(invocation, nameof(Equals)) && model.GetSymbolInfo(invocation).Symbol.ContainingAssembly.Name == "System.Private.CoreLib")
```

Comment:
This is definitely more precise, but you can make it slightly more performance by being as lazy as possible at evaluating the SM:

```csharp
        if (IsNameEqualTo(invocation, nameof(Equals)))
        {
            if (Language.Syntax.HasExactlyNArguments(invocation, 1)) // x.Equals(y)
            {
                return Language.Syntax.TryGetOperands(invocation, out var left, out _)
                    && HasInvocationValidOperands(left, GetArgumentExpression(invocation, 0))
                    && CheckIfSystemEquals();
            }
            if (Language.Syntax.HasExactlyNArguments(invocation, 2)) // Equals(x,y)
            {
                return HasInvocationValidOperands(GetArgumentExpression(invocation, 0), GetArgumentExpression(invocation, 1))
                    && CheckIfSystemEquals();
            }
        }
        return false;

        bool HasInvocationValidOperands(SyntaxNode first, SyntaxNode second) =>
            AreValidOperands(lambdaVariableName, first, second) || AreValidOperands(lambdaVariableName, second, first);

        bool CheckIfSystemEquals() =>
            model.GetSymbolInfo(invocation).Symbol.ContainingAssembly.Name == "System.Private.CoreLib";
```

Also, do you have a testcase that covers a custom-Equals method? (Should not raise)

Reply:
On a second note, this is quite troublesome. The problem is assemblies cannot be trusted, because of different framework versions and [reference assemblies](https://stackoverflow.com/questions/62128565/net-core-system-private-corelib-dll-vs-system-runtime).

Long story short, you cannot trust that `Equals` will _always_ be in `System.Private.CoreLib`.
Another possible check is the namespace:

```csharp
c.SemanticModel.GetSymbolInfo(thingy).Symbol.ContainingNamespace.Name == "System";
```

You can either use the namespace, find something similar or just drop this functionality, I am fine with all of the options.

Reply:
I moved the SM checks further as suggested and changed the check on the namespace. I think it can be valuable so I prefer to keep it instead of dropping the whole SM check. Regarding the UTs, there was one test with a custom Equal function with 3 parameters, I added another one with two for completeness. They should not raise for Contains but raise for Exists👍

Reply:
I agree, I think this will only raise if a user creates their one `System` namespace, which seems like very bad practise anyways.

Looks good, let's keep it.